src/Microsoft.AspNetCore.Pubternal.Analyzers/PubternalityAnalyzer.cs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Pubternal.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class PubternalityAnalyzer : DiagnosticAnalyzer
+    {
+        public PubternalityAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                PubturnalityDescriptors.PUB0001,
+                PubturnalityDescriptors.PUB0002
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(analysisContext =>
+            {
+                analysisContext.RegisterSymbolAction(symbolAnalysisContext => AnalyzeTypeUsage(symbolAnalysisContext), SymbolKind.Namespace);
+                analysisContext.RegisterSyntaxNodeAction(syntaxContext => AnalyzeTypeUsage(syntaxContext), SyntaxKind.IdentifierName);
+            });
+        }
+
+        private void AnalyzeTypeUsage(SymbolAnalysisContext context)
+        {
+            var ns = (INamespaceSymbol)context.Symbol;
+            if (ns.Name == "Internal")
+            {
+                return;
+            }
+
+            foreach (var namespaceOrTypeSymbol in ns.GetMembers())
+            {
+                if (namespaceOrTypeSymbol.IsType)
+                {
+                    CheckType((ITypeSymbol)namespaceOrTypeSymbol, context);
+                }
+            }
+        }
+
+        private void CheckType(ITypeSymbol typeSymbol, SymbolAnalysisContext context)
+        {
+            if (IsPrivate(typeSymbol) || IsPrivate(typeSymbol.ContainingType))
```

Comment:
Does this handle double-nested types?

```csharp
internal class Grandparent {
  public class Parent {
    public class Child {
    }
  }
}
```

Reply:
Yes https://github.com/aspnet/Common/pull/345/files#diff-fdfa35d2edfb0ca26c547defab430a66R82

Reply:
But, won't it consider a member in `Child` as "Public" even though it's contained within the `internal` `Grandparent` type? Or does this automatically walk up the ContainingType chain?

Reply:
Ah, no, we go top down starting from namespaces if the parent is private we won't even drill into it. https://github.com/aspnet/Common/pull/345/files#diff-fdfa35d2edfb0ca26c547defab430a66R56